tests/bases/test_ddp.py

```diff
@@ -138,30 +138,85 @@ def update(self, x):
         def compute(self):
             return self.x // self.c
 
+        def __repr__(self):
+            return f"DummyCatMetric(x={self.x}, c={self.c})"
+
     metric = DummyCatMetric()
     metric.persistent(True)
 
+    metric.to(f"cuda:{rank}")
+
+    def verify_metric(metric, i, world_size):
+        state_dict = metric.state_dict()
+        exp_sum = i * (i + 1) / 2
+        assert state_dict["x"] == exp_sum * world_size
+        assert metric.x == exp_sum * world_size
+        assert metric.c == (i + 1) * world_size
+        assert state_dict["c"] == metric.c
+
     steps = 5
     for i in range(steps):
+
+        if metric.is_synced:
+
+            with pytest.raises(MisconfigurationException, match="The Metric shouldn't be synced when performing"):
+                metric(i)
+
+            metric.unsync()
+
         metric(i)
-        state_dict = metric.state_dict()
 
-        exp_sum = i * (i + 1) / 2
-        assert state_dict["x"] == exp_sum * worldsize
-        assert metric.x == exp_sum
-        assert metric.c == (i + 1)
-        assert state_dict["c"] == metric.c * worldsize
+        verify_metric(metric, i, 1)
+
+        metric.sync()
+        assert metric.is_synced
+
+        with pytest.raises(MisconfigurationException, match="The Metric has already been synced."):
+            metric.sync()
+
+        verify_metric(metric, i, 2)
+
+        metric.unsync()
+        assert not metric.is_synced
+
+        with pytest.raises(MisconfigurationException, match="The Metric has already been un-synced."):
+            metric.unsync()
+
+        with metric.sync_context():
+            assert metric.is_synced
+            verify_metric(metric, i, 2)
+
+        with metric.sync_context(should_unsync=False):
+            assert metric.is_synced
+            verify_metric(metric, i, 2)
+
+        assert metric.is_synced
+
+        metric.unsync()
+
+        assert not metric.is_synced
+
+        metric.sync()
 
     def reload_state_dict(state_dict, expected_x, expected_c):
         metric = DummyCatMetric()
         metric.load_state_dict(state_dict)
         assert metric.x == expected_x
         assert metric.c == expected_c
 
-    with mock.patch.dict(os.environ, {"GLOBAL_RANK": str(rank)}):
-        reload_state_dict(deepcopy(state_dict), 20 if not rank else 0, 10 if not rank else 0)
+    reload_state_dict(deepcopy(metric.state_dict()), 20, 10)
+
+    metric.unsync()
+    reload_state_dict(deepcopy(metric.state_dict()), 10, 5)
+
+    metric.cpu()
+    metric.sync()
+
+    torch.save(metric.state_dict(), os.path.join(tmpdir, 'weights.pt'))
 
-    reload_state_dict(deepcopy(state_dict), 20, 10)
+    metric.unsync()
+    with metric.sync_context():
+        torch.save(metric.state_dict(), os.path.join(tmpdir, 'weights.pt'))
```

Comment:
could lead to a flaky test if saving on both ranks at the same time :)

Reply:
Good catch !

Reply:
Done !